### PR TITLE
Support all iceberg types in python library

### PR DESCRIPTION
**Pull request by jun-he**
For https://github.com/apache/iceberg/issues/3216
Note: I collapse type.py and types.py in the python_legacy to the new library, which is similar to existing Java implementation. Please take a look at it. We can use it as an example to discuss how to move towards for pythonic refactoring.

### CODE REVIEW

File: python/src/iceberg/types.py [hunk 1]
```diff
@@ -0,0 +1,824 @@
...
+    DECIMAL = {"java_class": "BigDecimal.class", "python_class": Decimal, "id": 13}
+    STRUCT = {"java_class": "Void.class", "python_class": None, "id": 14}
+    LIST = {"java_class": "Void.class", "python_class": None, "id": 15}
+    MAP = {"java_class": "Void.class", "python_class": None, "id": 16}
+
+
+class Type(object):
+    length: int
+    scale: int
+    precision: int
```
Comment (rdblue) [inline python/src/iceberg/types.py:46]: Do these need to be here or on the specific types, fixed and decimal?

File: python/src/iceberg/types.py [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+        pass
+
+    def type_id(self):
+        pass
+
+    def is_primitive_type(self):
+        return False
+
+    def as_primitive_type(self):
+        raise ValueError("Not a primitive type: " + self)
```
Comment (rdblue) [inline python/src/iceberg/types.py:58]: I think these `as_` and `is_` methods are valuable if we have type annotations. Is the plan to add them later?

**Comment by samredai** [comment]
@jun-he do you think you could give me access to this branch `jun-he:jun/add-types` in your forked repo? I have a commit queued up to add test coverage to pass CI and figured we can do that in the same PR. I also added some of the additional linters/autoformatting we discussed.

**Comment by samredai** [comment]
Should we add the auto formatting stuff in this PR to prevent having to do a huge auto format later? (It's also fine to do in a follow up PR). I was thinking something like:

```ini
[testenv:format]
description = reformat all source code
basepython = python3
deps =
    black
    isort
    flake8
skip_install = true
commands =
    isort --recursive --project iceberg --profile black setup.py src tests
    black setup.py src tests
    flake8 setup.py src tests

[testenv:linters]
basepython = python3
skip_install = true
deps =
    .
    {[testenv:isort]deps}
    {[testenv:black]deps}
    {[testenv:flake8]deps}
    {[testenv:bandit]deps}
    {[testenv:mypy]deps}
commands =
    {[testenv:isort]deps}
    {[testenv:black]deps}
    {[testenv:flake8]commands}
    {[testenv:bandit]commands}
    {[testenv:mypy]commands}

[testenv:isort]
basepython = python3
skip_install = true
deps =
    isort
commands =
    isort --recursive --project iceberg --profile black --check-only setup.py src tests

[testenv:black]
basepython = python3
skip_install = true
deps =
    black
commands =
    black --check --diff src setup.py tests
```

**Comment by rdblue** [comment]
@samredai, if we can format the code here, let's do that. But I wouldn't say let's update the config for formatting, since that's a separate change.

**Pull request merged**